### PR TITLE
[core] Implement filter string parsing to Narwhals expressions

### LINKED ISSUE #58: [core] Implement filter string parser to Narwhals expressions
## Description
Implement a parser that converts filter strings into Narwhals expressions (`IntoExpr`). This will enable users to specify filters as strings that can be parsed and applied to Narwhals DataFrames/LazyFrames, remaining agnostic to the underlying storage backend.

The main goal is to convert string expressions (like SQL WHERE clauses) into Narwhals expressions that can be applied to any Narwhals-compatible backend (Polars, Pandas, PyArrow, etc.).

This is a prerequisite for the metadata query format feature, where filters will be specified as part of query strings.

## Research Summary
Investigation of existing libraries found:

### Option 1: SQLGlot (Recommended)
- No-dependency SQL parser supporting 31 dialects
- Can parse SQL WHERE expressions into AST
- Provides rich API for AST manipulation and traversal
- Example: `sqlglot.parse_one("age > 25 AND status = 'active'")`
- Can transform parsed AST into Narwhals expressions

### Option 2: Build Custom Parser
- Use pyparsing or lark for robustness
- Full control over syntax and semantics
- More maintenance overhead

### Note on existing libraries:
- **Narwhals**: No built-in string-to-expression parsing (which is why we need this)
- **PyArrow**: Has Expression API but requires programmatic construction, no string parsing

## Acceptance Criteria
- [ ] Implement parser using SQLGlot to convert filter strings to Narwhals expressions
- [ ] Support comparison operators: `=`, `!=`, `>`, `<`, `>=`, `<=`
- [ ] Support logical operators: `AND`, `OR`, `NOT`
- [ ] Support field references and literal values (strings, numbers, booleans, null)
- [ ] Support parentheses for grouping expressions
- [ ] Convert parsed AST to Narwhals `IntoExpr` types
- [ ] Ensure expressions work across all Narwhals-supported backends
- [ ] Add comprehensive unit tests covering all operators and edge cases
- [ ] Document supported SQL WHERE clause syntax
- [ ] Handle parse errors gracefully with informative error messages

## Implementation Notes
- Use SQLGlot to parse SQL WHERE clause syntax into AST
- Transform SQLGlot AST nodes into corresponding Narwhals expressions
- The output must be backend-agnostic Narwhals expressions
- Reference: https://narwhals-dev.github.io/narwhals/api-reference/typing/#narwhals.typing.IntoExpr
- This parser will be used by metadata stores to filter data regardless of backend

## Example Usage
```python
import narwhals as nw

# Parse simple comparison
expr = parse_filter_string("age > 25")
# Returns: nw.col("age") > 25

# Parse compound expression (SQL WHERE syntax)
expr = parse_filter_string("age > 25 AND status = 'active'")
# Returns: (nw.col("age") > 25) & (nw.col("status") == "active")

# Parse with parentheses
expr = parse_filter_string("(age > 25 OR age < 18) AND status != 'deleted'")
```

### PR DESCRIPTION
**Pull request by geoHeil**
resolves #58

### CODE REVIEW

File: src/metaxy/models/filter_expression.py [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+    type: Literal["operand"] = "operand"
+    operand: FilterOperandType
+
+    def to_expr(self) -> nw.Expr:
+        return self.operand.to_expr()
+
+
+FilterExpressionType = Annotated[
+    ComparisonExpression | LogicalExpression | NotExpression | OperandExpression,
+    Field(discriminator="type"),
```
Comment (geoHeil) [inline src/metaxy/models/filter_expression.py]: do we need any input validation /  SQL injection handlning?

Comment (geoHeil) [reply src/metaxy/models/filter_expression.py]: also given that internal (trusted) DE would control this

File: src/metaxy/models/filter_expression.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+
+    expression: FilterExpressionType
+    source: str | None = None
+
+    def to_expr(self) -> nw.Expr:
+        """Convert the stored expression into a Narwhals ``Expr``."""
+        return self.expression.to_expr()
+
+    @classmethod
+    def from_string(cls, filter_string: str) -> NarwhalsFilter:
```
Comment (danielgafni) [inline src/metaxy/models/filter_expression.py]: this should be a Pydantic validator instead

File: src/metaxy/models/filter_expression.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+"""
+
+from __future__ import annotations
+
+from enum import Enum
+from typing import Annotated, Any, ClassVar, Literal
+
+import narwhals as nw
+from pydantic import Field, field_validator
+from sqlglot import exp, parse_one
```
Comment (danielgafni) [inline src/metaxy/models/filter_expression.py]: maybe keep it as `import sqlglot`, this is the most important lib in this file and it should be clear where it is used vs our functions

**Review by danielgafni — CHANGES_REQUESTED** [review]
I'm happy this works (clear from tests) but something is telling me it is overly complicated. I am sure `sqlglot` has a ready API for this. Could you please ensure this is not the case before rolling out something like this? 

Best if you ask them on github or somewhere else if unsure.

File: src/metaxy/models/filter_expression.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+            if isinstance(maybe_source, str) and "expression" not in data:
+                expression = _parse_to_expression(maybe_source)
+                updated = dict(data)
+                updated["expression"] = expression
+                return updated
+
+        return data
+
+    @field_serializer("expression")
+    def _serialize_expression(self, expression: sqlglot.exp.Expression) -> str:
```
Comment (danielgafni) [inline src/metaxy/models/filter_expression.py]: Ok I guess it isn't a Pydantic model then

File: src/metaxy/models/filter_expression.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+        msg = f"Failed to parse filter string: {exc}"
+        raise FilterParseError(msg) from exc
+
+    if parsed is None:
+        raise FilterParseError("Failed to parse filter string into an expression.")
+
+    return parsed
+
+
+def _expression_to_narwhals(node: sqlglot.exp.Expression) -> nw.Expr:
```
Comment (danielgafni) [inline src/metaxy/models/filter_expression.py]: Rename to "swlglot expression to narwhals" or smth like this

File: src/metaxy/models/filter_expression.py [hunk 6]
```diff
@@ -0,0 +1,337 @@
...
+    expr: nw.Expr
+    is_literal: bool
+    literal_value: LiteralValue
+    is_column: bool
+
+
+class NarwhalsFilter(FrozenBaseModel):
+    """Pydantic model for serializable Narwhals filter expressions."""
+
+    expression: sqlglot.exp.Expression
```
[the remark this answers is not in the document]
Comment (geoHeil) [reply src/metaxy/models/filter_expression.py]: yes FrozenBaseModel is 

```python
import pydantic


class FrozenBaseModel(pydantic.BaseModel):
    # config class is deprecated
    model_config = pydantic.ConfigDict(frozen=True)

```

File: src/metaxy/models/filter_expression.py [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+    @field_serializer("expression")
+    def _serialize_expression(self, expression: sqlglot.exp.Expression) -> str:
+        return expression.sql()
+
+    def to_expr(self) -> nw.Expr:
+        """Convert the stored expression into a Narwhals ``Expr``."""
+        return _expression_to_narwhals(self.expression)
+
+    @classmethod
+    def from_string(cls, filter_string: str) -> NarwhalsFilter:
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply src/metaxy/models/filter_expression.py]: But how is it different from `.model_validate`?

File: src/metaxy/models/filter_expression.py [hunk 8]
```diff
@@ -0,0 +1,337 @@
...
+from metaxy.models.bases import FrozenBaseModel
+
+LiteralValue = bool | int | float | str | None
+
+
+class FilterParseError(ValueError):
+    """Raised when a filter string cannot be parsed into a supported expression."""
+
+
+class ComparisonOperator(str, Enum):
```
[the remark this answers is not in the document]
Comment (geoHeil) [reply src/metaxy/models/filter_expression.py]: I could not find it so far. 


But would we be able to untyped on the AST level pass the sqlglot over to narwhals? instead

File: src/metaxy/models/filter_expression.py [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+        raise FilterParseError("Filter string cannot be empty.")
+
+    try:
+        parsed = sqlglot.parse_one(filter_string)
+    except ParseError as exc:
+        msg = f"Failed to parse filter string: {exc}"
+        raise FilterParseError(msg) from exc
+
+    if parsed is None:
+        raise FilterParseError("Failed to parse filter string into an expression.")
```
[the remark this answers is not in the document]
Comment (geoHeil) [reply src/metaxy/models/filter_expression.py]: separate issue there for loguru; fstring implemented

**Review by danielgafni — CHANGES_REQUESTED** [review]
Wow, looking much cleaner now! 

I think it's very close. 
I left 2 minor comments. 

Also, I would like to see more tests with more complicated expressions: make it obvious that parsing works correctly even for more nested expressions with parentheses, etc. 

Maybe just 1-2 more tests.

File: docs/reference/api/types.md [hunk 10]
```diff
@@ -13,3 +13,25 @@ A few types used in Metaxy here and there.
...
+They are backed by SQLGlot and understand the following subset of SQL `WHERE` syntax:
+
+- Comparisons: `=`, `!=`, `>`, `<`, `>=`, `<=`
+- Logical operators: `AND`, `OR`, `NOT`
+- Parentheses for grouping
+- Column references (identifiers or dotted paths)
+- Literals: strings (`'value'`), numbers, booleans (`TRUE`/`FALSE`), and `NULL`
+- Implicit boolean columns (e.g., `NOT is_active`)
+
+::: metaxy.models.filter_expression.NarwhalsFilter
```
[the remark this answers is not in the document]
Comment (geoHeil) [reply docs/reference/api/types.md]: moved down

File: src/metaxy/models/filter_expression.py [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+    """Parse a SQL WHERE-like string into a Narwhals expression.
+
+    The parser uses SQLGlot under the hood and understands SQL WHERE
+    clauses composed of comparison operators, logical operators, parentheses,
+    dotted identifiers, and literal values (strings, numbers, booleans, ``NULL``).
+    It returns a backend-agnostic Narwhals expression that can be fed directly
+    into ``LazyFrame.filter`` for any Narwhals-compatible backend (Polars,
+    pandas, PyArrow, DuckDB, ...).
+
+    Examples
```
Comment (danielgafni) [inline src/metaxy/models/filter_expression.py]: Looks like the format here isn't mkdocs

Are you looking at docs locally? 

`just docs-serve`

File: src/metaxy/migrations/detector.py [hunk 12]
```diff
@@ -155,6 +155,8 @@ def detect_migration(
 
     yaml_path = migrations_dir / f"{migration_id}.yaml"
     migration_yaml = {
+        "migration_type": migration.migration_type,
```
Comment (danielgafni) [inline src/metaxy/migrations/detector.py]: What's these changes about?

File: docs/reference/api/types.md [hunk 13]
```diff
@@ -13,3 +13,30 @@ A few types used in Metaxy here and there.
 ::: metaxy.models.types.SnapshotPushResult
 
 ::: metaxy.IDColumns
+
+## Narwhals Filter Expressions
+
+Narwhals filter expressions let you serialize and deserialize backend-agnostic predicates.
+They are backed by SQLGlot and understand the following subset of SQL `WHERE` syntax:
+
+- Comparisons: `=`, `!=`, `>`, `<`, `>=`, `<=`
```
Comment (danielgafni) [inline docs/reference/api/types.md]: I think this list is missing new lines - are you rendering the docs for yourself?

**Review by danielgafni — CHANGES_REQUESTED** [review]
Great, I like the main diff now, but we should stop irrelevant changes (e.g. migrations code) and fix the docs. Render them locally to make sure the format is correct.

File: src/metaxy/migrations/detector.py (hunk 12 above)
Comment (geoHeil) [reply src/metaxy/migrations/detector.py]: was required after the rebase from main to make sure the tests do not fail

Comment (geoHeil) [reply src/metaxy/migrations/detector.py]: actually not only for this branch here

**Review by danielgafni — CHANGES_REQUESTED** [review]

File: src/metaxy/models/filter_expression.py (hunk 11 above)
Comment (danielgafni) [reply src/metaxy/models/filter_expression.py]: You have to add it, see below

Comment (geoHeil) [reply src/metaxy/models/filter_expression.py]: fixed now

File: src/metaxy/migrations/detector.py [hunk 14]
```diff
@@ -37,7 +37,7 @@ def detect_migration(
     Returns:
         DiffMigration if changes detected and written, None otherwise
 
-    Example:
+    Examples:
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply src/metaxy/migrations/detector.py]: Yep, confirmed `Examples` isn't a valid keyword. 

`Example` is. 
image

[the remark this answers is not in the document]
Comment (geoHeil) [reply src/metaxy/migrations/detector.py]: dropped

File: tests/models/test_filter_expression.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+            [2, 4],
+        ),
+        (
+            "NOT (status = 'deleted' OR deleted_at != NULL) "
+            "AND (age < 30 OR NOT is_active)",
+            [1, 2],
+        ),
+    ],
+)
+def test_parse_deeply_nested_expressions(
```
[the remark this answers is not in the document]
Comment (danielgafni) [reply tests/models/test_filter_expression.py]: nice! much cleaner

File: src/metaxy/models/filter_expression.py [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+    """Parse a SQL WHERE-like string into a Narwhals expression.
+
+    The parser uses SQLGlot under the hood and understands SQL WHERE
+    clauses composed of comparison operators, logical operators, parentheses,
+    dotted identifiers, and literal values (strings, numbers, booleans, ``NULL``).
+    It returns a backend-agnostic Narwhals expression that can be fed directly
+    into ``LazyFrame.filter`` for any Narwhals-compatible backend (Polars,
+    pandas, PyArrow, DuckDB, ...).
+
+    Examples:
```
[the remark this answers is not in the document]
Comment (geoHeil) [reply src/metaxy/models/filter_expression.py]: done

[the remark this answers is not in the document]
Comment (danielgafni) [reply src/metaxy/models/filter_expression.py]: I still see `Examples` here. 

Also, perhaps one example with some nested expression is enough? I don't think additional examples add any value, it should already be obvious this works with any kind of filtering expression.

**Review by danielgafni — CHANGES_REQUESTED** [review]

File: docs/reference/api/types.md [hunk 17]
```diff
@@ -13,3 +13,36 @@ A few types used in Metaxy here and there.
...
+# Parse a SQL WHERE clause into a backend-agnostic Narwhals expression
+expr = parse_filter_string("(age > 25 OR age < 18) AND status != 'deleted'")
+
+# Apply the parsed expression to any Narwhals LazyFrame
+filtered = lazy_frame.filter(expr)
+```
+
+---
+
+::: metaxy.models.filter_expression.NarwhalsFilter
```
Comment (danielgafni) [inline docs/reference/api/types.md]: I just realized this isn't a public-facing API at all, maybe remove it from API docs?

**Review by danielgafni — CHANGES_REQUESTED** [review]

**Comment by danielgafni** [comment]
If you are fine with that - sure
